bug fix in LP9

WegrzynAB opened the pull request:
Error in Line 27 while using models with coupling constraints:

LPproblem.A = [model.S,model.E;model.C,model.D]; is bigger in models with coupling constraints than model.S
while sparse(m,np) takes its dimension m from model.S only
this creates inequality in dimensionality in Aeq = [LPproblem.A, sparse(m,np)];

proposed fix:
Aeq = [LPproblem.A, sparse(m2,np)]; creates sparse of dimensionality equal to LPproblem.A 

**I hereby confirm that I have:**

- [x] Tested my code on my own machine
- [x] Followed the guidelines in the [Contributing Guide](https://opencobra.github.io/cobratoolbox/docs/contributing.html)
- [x] Selected `develop` as a target branch (top left drop-down menu)

*(Note: You may replace [ ] with [X] to check the box)*

laurentheirendt commented:
@WegrzynAB, great! Is there any way you could add a quick test for this?

WegrzynAB commented:
Sure, I will work on it.

WegrzynAB commented:
@laurentheirendt I wrote a test where I check the consistency of the results from the original LP9.m script and the new version of LP9.m using E.coli model (model without coupling constraints). [See test files](https://github.com/WegrzynAB/cobratoolbox/tree/bugFix_LP9/test/additionalTests/LP9test) 
Is this something that you had in mind?

laurentheirendt commented:
May you remove `test/additionalTests/LP9test/testLP9.m~`? This is a hidden file actually. 👍

WegrzynAB commented:
done!

laurentheirendt commented:
I am not sure I actually understand your testing method here. 

It seems as if the source code is copied for generating reference data. That is not such a good idea. It might be best to simply save the reference data in a `.mat` file generated with the current version of the function that you want to test.

The test also has to be in `verifiedTests` for now, otherwise it is not run as part of the test suite on ARTENOLIS.

laurentheirendt commented on test/additionalTests/LP9test/testLP9.m:
```diff
@@ -0,0 +1,87 @@
+        end
+        K = intersect(J, find(V >= 0.99*epsilon));
+        % create reference data
+        V_ref = LP9(K, P, model, LPproblem, epsilon);
```
`V_ref` should be loaded from a `.mat` file. Otherwise, we will have 2 versions of the code, one version in `src`, and one version in `test` (and named differently).

laurentheirendt commented on test/additionalTests/LP9test/testLP9.m:
```diff
@@ -0,0 +1,87 @@
+        % created test data for new version
+        V_test = LP9_test(K, P, model, LPproblem, epsilon);
+    end
+    if V_ref == V_test
```
This should be an `assert`.

laurentheirendt commented on test/additionalTests/LP9test/testLP9.m:
```diff
@@ -0,0 +1,87 @@
+[nMets,nRxns] = size(model.S);
+
+for k = 1:length(solversPkgs.LP)
+    fprintf(' -- Running <testFile> using the solver interface: %s ... ', solversPkgs.LP{k});
```
This has to be adapted to the test at hand.

laurentheirendt commented on test/additionalTests/LP9test/testLP9.m:
```diff
@@ -0,0 +1,87 @@
+global CBTDIR
+
+% define the features required to run the test
+requiredToolboxes = {};
```
This is not needed here.

laurentheirendt commented on test/additionalTests/LP9test/testLP9.m:
```diff
@@ -0,0 +1,87 @@
+% define the features required to run the test
+requiredToolboxes = {};
+
+requiredSolvers = {'glpk'};
```
Eventually, it might be best to use `gurobi` for the test.

WegrzynAB replied:
currently, I do not have Gurobi installed so I cannot test with it

WegrzynAB replied:
Ok, I installed Gurobi and edited the test to work with it.

laurentheirendt commented on test/additionalTests/LP9test/LP9_test.m:
```diff
@@ -0,0 +1,66 @@
+function V = LP9_test(K, P, model, LPproblem, epsilon)
```
This function here should be removed.

WegrzynAB commented:
I have followed your comments and changed the test script, created the .mat files with options, and reference data, changed solver to gurobi, and moved test folder to [verifiedTests](https://github.com/WegrzynAB/cobratoolbox/tree/bugFix_LP9/test/verifiedTests).

laurentheirendt commented:
Great! 👍 Actually, the test in `test/verifiedTests` should be in the appropriate subfolder.

Now, the test makes a lot more sense ...

WegrzynAB commented:
Hi, I have moved the testLP9 to testFASTCORE as it is part of the fastcore method; I hope that now everything is in order.

laurentheirendt commented:
@WegrzynAB, the tests do not pass. May you have a look? Thanks.
```
18:01:44 The reasons were as follows:
18:01:44 ------------------------------------------------
18:01:44 testLP9.m:
18:01:44 Error using testLP9 (line 45)
18:01:44 
18:01:44 Results (V) are not consistent between old and new version of LP9
18:01:44 
18:01:44 ------------------------------------------------
```

WegrzynAB commented:
On my local PC (Ubuntu 18.04) this test passes without a problem. I think the issue here is the small variability in the results depending on the platform and PC setup (rounding errors etc.). Which is why I originally intended to create the reference set in the same run as the test. I need to figure out how to relax the test to account for those small variations. I will try to run it locally on a different platform (Linux vs Mac) and see if I can reproduce the error you're getting.
`>> testLP9
 -- Running testLP9.m using the solver interface: gurobi ... 
Done.`

laurentheirendt commented on test/verifiedTests/analysis/testFASTCORE/LP9test/testLP9.m:
```diff
@@ -0,0 +1,51 @@
+        % created test data for new version
+        V_test = LP9(options.K, options.P, model, options.LPproblem, options.epsilon);
+    end
+    assert(isequal(reference_V,V_test), '\nResults (V) are not consistent between old and new version of LP9\n')
```
Instead of using here `isequal`, you can set a `tolerance` such as explained [here](https://opencobra.github.io/cobratoolbox/stable/contributing.html#test-if-an-output-is-correct).

rmtfleming commented:
There are probably alternate optimal solutions. Just test that the value of
the objective is the same, rather than the vector.

On Mon, 11 Mar 2019 at 16:18, Laurent Heirendt <notifications@github.com>
wrote:

> *@laurentheirendt* commented on this pull request.
> ------------------------------
>
> In test/verifiedTests/analysis/testFASTCORE/LP9test/testLP9.m
> <https://github.com/opencobra/cobratoolbox/pull/1435#discussion_r264282695>
> :
>
> > +load('options.mat');
> +
> +% load reference data
> +load('reference_V.mat');
> +
> +% load
> +for k = 1:length(solversPkgs.LP)
> +    fprintf(' -- Running testLP9.m using the solver interface: %s ... ', solversPkgs.LP{k});
> +
> +    solverLPOK = changeCobraSolver(solversPkgs.LP{k}, 'LP', 0);
> +
> +    if solverLPOK
> +        % created test data for new version
> +        V_test = LP9(options.K, options.P, model, options.LPproblem, options.epsilon);
> +    end
> +    assert(isequal(reference_V,V_test), '\nResults (V) are not consistent between old and new version of LP9\n')
>
> Instead of using here isequal, you can set a tolerance such as explained
> here
> <https://opencobra.github.io/cobratoolbox/stable/contributing.html#test-if-an-output-is-correct>
> .
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/opencobra/cobratoolbox/pull/1435#pullrequestreview-212899616>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACDCOoYZaBccOspVBVuhk3u_7kmaMIVhks5vVnO-gaJpZM4bXgU4>
> .
>


-- 
--
Mr. Ronan MT Fleming B.V.M.S. Dip. Math. Ph.D.
----------------------------------------------------------------------------
Assistant Professor,
Division of Systems Biomedicine and Pharmacology,
Leiden Academic Centre for Drug Research,
Faculty of Science,
Leiden University.
https://www.universiteitleiden.nl/en/staffmembers/ronan-fleming
&
H2020 Project Coordinator
Systems Medicine of Mitochondrial Parkinson’s Disease
http://sysmedpd.eu
----------------------------------------------------------------------------
Peer-reviewed publications: https://goo.gl/FZPG23
Mobile:  +353 873 413 072
Skype: ronan.fleming
----------------------------------------------------------------------------
(This message is confidential and may contain privileged information. It is
intended for the named recipient only. If you receive it in error please
notify me and permanently delete the original message and any copies.)

WegrzynAB commented:
Hi, Laurent, can you run the test and see if it works properly this time? Again, I tested it locally and it works.

laurentheirendt commented:
May you revert the commit `db6f181`  and may you rename the reference data file according to the [guidelines](https://opencobra.github.io/cobratoolbox/stable/contributing.html)? The tests are run on ARTENOLIS automatically. 

Thanks 👍
